src/battle/battle.cpp

```diff
@@ -1,50 +1,117 @@
 #include "battle.h"
+#include "foe_attack.h"
+#include "main_character_attack.h"
 #include "../utils/utils.h"
 #include "../creature/creature.h"
 #include "../creature/main_character/main_character.h"
 #include <iostream>
 #include <vector>
+#include <algorithm>
 
 void Battle(MainCharacter* main_character, std::vector<Creature*>& foes)
 {
-	int victories = 0;
-	int losses = 0;
 	std::cout << "Battle starts!" << std::endl;
 
-	main_character->PrintStats();
+	// All the battle participants (including main character)
+	// Contains pairs of type { foe/main_character, initiative_roll }
+	std::vector<std::pair<Creature*, int>> combatants(foes.size() + 1,
+		{foes[0], 0}); // Add +1 to accommodate for the main character
 
-	for (const auto& foe : foes)
+	/**
+	 * Roll initiative for main_character and foes,
+	 * and initialize the combatants list
+	 */
+
+	int main_character_initiative_roll =
+		utils::RollDice(1, 20) + (main_character->get_ability_score(1) - 10) / 2;
+	std::cout << "Main character rolled initiative "
+			  << main_character_initiative_roll << std::endl;
+
+	combatants[0] = {main_character, main_character_initiative_roll};
+
+	for (size_t i = 0; i < foes.size(); ++i)
 	{
-		std::cout << "Stats for " << foe->get_creature_type() << ":" << std::endl;
-		foe->PrintStats();
+		int foe_initiative_roll =
+			utils::RollDice(1, 20) + (foes[i]->get_ability_score(1) - 10) / 2;
+		std::cout << "Foe rolled initiative " << foe_initiative_roll << std::endl;
 
-		int foe_roll = utils::RollDice(1, 20) + foe->get_ability_score(0);
-		std::cout << "foe rolled " << foe_roll << std::endl;
+		combatants[i + 1] = {foes[i], foe_initiative_roll};
+	}
 
-		int main_character_roll =
-			utils::RollDice(1, 20) + main_character->get_ability_score(0);
-		std::cout << "main character rolled " << main_character_roll << std::endl;
+	// Sort the combatants list in descending order of initiative rolls
+	const auto compareCombatantsInitiative = [](std::pair<Creature*, int> a,
+												 std::pair<Creature*, int> b) -> bool
+	{ return (a.second > b.second); };
```

Comment:
If we wanna be completely rules-compliant, we should consider the case where they have the same initiative roll. Then, first goes the one with the highest dexterity score.